File: ZebraSftpMpBodyPublisher.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Net.Http;
using System.Linq;

public class ZebraSftpMpBodyPublisher
{
    private List<String> serialNumberList = new List<String>(); // At least to have one printer SN
    private string zplFilePath = null; // Must have one zplFilePath to build the multipart request body.

    public ZebraSftpMpBodyPublisher()
    {
    }

    public MultipartFormDataContent Build()
    {
        if (serialNumberList.Count == 0) {
            throw new ArgumentOutOfRangeException("Must have at least one printer serial number in order to build multipart form data.");
        }

        if (zplFilePath == null) {
            throw new ArgumentNullException("Must have a ZPL file in order to build multipart form data.");
        } else if (!File.Exists(zplFilePath)) {
            throw new ArgumentException("The specified ZPL file does not exist.");
        }

        // Compose the multipart form data content
        var form = new MultipartFormDataContent();

        // Remove the quotes around the boundary
        var boundary = form.Headers.ContentType.Parameters.First(o => o.Name == "boundary");
        boundary.Value = boundary.Value.Replace("\"", String.Empty);

        // Add printer serial number(s)
        foreach (string sn in serialNumberList) {
            form.Add(new StringContent(sn), "\"sn\""); // Extra double quotes around "sn" are needed
        }

        FileStream zplFileStream = new FileStream(zplFilePath, FileMode.Open);
        form.Add(new StreamContent(zplFileStream), "zpl_file", Path.GetFileName(zplFilePath));

        return form;
    }

    // Add a printer serial number. You can add the serial number multiple times,
    // or add multiple serial numbers.
    public ZebraSftpMpBodyPublisher AddSn(string serialNumber)
    {
        serialNumberList.Add(serialNumber);
        return this;
    }

    // You can have only one ZPL file at a time.
    public ZebraSftpMpBodyPublisher SetZplFilePath(string zplFilePath)
    {
        this.zplFilePath = zplFilePath;
        return this;
    }
}

File: ZebraSftpMpBodyPublisher.java
package com.zebra.sendfiletoprinter.sample;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.net.http.HttpRequest;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.UUID;

public class ZebraSftpMpBodyPublisher {

    private String boundary = UUID.randomUUID().toString();

    private List<String> serialNumberList = new ArrayList<String>(); // At least to have one printer SN
    private Path zplFilePath = null; // Must have one zplFilePath to build the multipart request body.

    public HttpRequest.BodyPublisher build() throws IOException {
        if (serialNumberList.size() == 0) {
            throw new IllegalStateException("Must have at least one printer serial number to build multipart request body.");
        }
        if (zplFilePath == null) {
            throw new IllegalStateException("Must have a ZPL file to build multipart request body.");
        }
        return HttpRequest.BodyPublishers.ofByteArrays(ZebraSftpPartsIterator::new);
    }

    public String getBoundary() {
        return boundary;
    }

    // Add a printer serial number. You can have more than one serial numbers.
    public ZebraSftpMpBodyPublisher addSn(String serialNumber) {
        serialNumberList.add(serialNumber);
        return this;
    }

    public ZebraSftpMpBodyPublisher setZplFilePath(Path zplFilePath) {
        this.zplFilePath = zplFilePath;
        return this;
    }

    class ZebraSftpPartsIterator implements Iterator<byte[]> {

        private Iterator<String> snIter;
        private InputStream zplFileInput;

        private boolean zpl_file_content_done = false;
        private boolean final_boundary_done = false;

        private boolean done;
        private byte[] next;

        ZebraSftpPartsIterator() {
            snIter = serialNumberList.iterator();
        }

        @Override
        public boolean hasNext() {
            if (done)
                return false;
            if (next != null)
                return true;
            try {
                next = calcuNextPart();
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
            if (next == null) {
                done = true;
                return false;
            }
            return true;
        }

        @Override
        public byte[] next() {
            if (!hasNext())
                throw new NoSuchElementException();
            byte[] tmp = next;
            next = null;
            return tmp;
        }

        private byte[] calcuNextPart() throws IOException {

            // First, get the parts for all the serial numbers.
            if (snIter.hasNext()) {
                String snBody = "--" + boundary + "\r\n" + "Content-Disposition: form-data; name=\"sn\"" + "\r\n\r\n" + snIter.next() + "\r\n";
                return snBody.getBytes(StandardCharsets.UTF_8);

            } else if (!zpl_file_content_done) {
                // Done with SN. Then get the part for ZPL file.
                if (zplFileInput == null) {
                    // Open the file input stream
                    zplFileInput = Files.newInputStream(zplFilePath);

                    // Create the header for ZPL file part
                    String contentType = Files.probeContentType(zplFilePath);
                    String zplFileBody = "--" + boundary + "\r\n" + "Content-Disposition: form-data; name=\"zpl_file\"; filename=\"" + zplFilePath.getFileName().toString() + "\"" + "\r\n"
                            + "Content-Type: " + contentType + "\r\n\r\n";
                    return zplFileBody.getBytes(StandardCharsets.UTF_8);
                } else {
                    // Read the file content
                    byte[] bytes = zplFileInput.readAllBytes();
                    zplFileInput.close();
                    zplFileInput = null;
                    byte[] actual = new byte[bytes.length + 2];
                    byte[] newline = "\r\n".getBytes(StandardCharsets.UTF_8); // Ending for the file content part.
                    System.arraycopy(bytes, 0, actual, 0, bytes.length);
                    System.arraycopy(newline, 0, actual, bytes.length, newline.length);
                    zpl_file_content_done = true;
                    return actual;
                }
            } else if (!final_boundary_done) {
                // Add the final boundary
                final_boundary_done = true;
                String finalBoundary = "--" + boundary + "--";
                return finalBoundary.getBytes(StandardCharsets.UTF_8);
            } else {
                // All done
                return null;
            }
        }
    }
}
